src/ballots/vote.cpp

```diff
@@ -0,0 +1,121 @@
+#include <ballots/vote.hpp>
+#include <dao.hpp>
+#include <common.hpp>
+#include <document_graph/edge.hpp>
+#include <member.hpp>
+#include <util.hpp>
+#include <trail.hpp>
+
+#define CONTENT_GROUP_LABEL_VOTE "vote"
+
+namespace hypha 
+{
+    Vote::Vote(hypha::dao& dao, const eosio::checksum256& hash)
+    : TypedDocument(dao, hash)
+    {
+
+    }
+    
+    Vote::Vote(
+        hypha::dao& dao, 
+        const eosio::name voter, 
+        std::string vote, 
+        Document& proposal
+    )
+    : TypedDocument(dao)
+    {
+        // Could be replaced by i.e. proposal.hasVote(vote) when proposal is no longer just a "Document"
+        // the goal is to have an easier API
+        proposal.getContentWrapper().getOrFail(BALLOT_OPTIONS, vote, "Invalid vote");
+        
+        eosio::check(
+            Edge::exists(dao.get_self(), getRoot(dao.get_self()), proposal.getHash(), common::PROPOSAL),
+            "Only allowed to vote active proposals"
+        );
+
+        std::vector<Edge> votes = dao.getGraph().getEdgesFrom(proposal.getHash(), common::VOTE);
+        for (auto votesIt = votes.begin(); votesIt != votes.end(); ++votesIt) {
+            if (votesIt->getCreator() == voter) {
+                eosio::checksum256 voterHash = Member::calcHash(voter);
+                Document voteDocument(dao.get_self(), votesIt->getToNode());
+
+                // Already voted, erase edges and allow to vote again.
+                Edge::get(dao.get_self(), voterHash, voteDocument.getHash(), common::VOTE).erase();
+                Edge::get(dao.get_self(), proposal.getHash(), voteDocument.getHash(), common::VOTE).erase();
+                Edge::get(dao.get_self(), voteDocument.getHash(), voterHash, common::OWNED_BY).erase();
+                Edge::get(dao.get_self(), voteDocument.getHash(), proposal.getHash(), common::VOTE_ON).erase();
```

Comment:
I think this could be simplified with this function https://github.com/hypha-dao/document-graph/blob/2f1ed8d124e2f112dd4dc01ef0055a59df4d6d3c/src/document_graph/document_graph.cpp#L87 
to `dao.getGraph().removeEdges(voteDocument.getHash())` unless there are more edges for the Vote document you would like to keep. Also adding a `break;` after this wouldn't harm :D

Reply:
> I think this could be simplified with this function https://github.com/hypha-dao/document-graph/blob/2f1ed8d124e2f112dd4dc01ef0055a59df4d6d3c/src/document_graph/document_graph.cpp#L87
> to `dao.getGraph().removeEdges(voteDocument.getHash())` unless there are other edges for the Vote document you would like to keep.

There could be other edges pointing to this document, the "Vote" document is too generic that chances are that it willl be resued for the same user (at least in this point).

> Also adding a break; after this wouldn't harm :D

+1